File: backend/app/schemas/api_key.py
"""
API密钥相关的 Pydantic Schemas
"""
from pydantic import BaseModel, Field, validator
from typing import Optional, List
from datetime import datetime
from enum import Enum


class ApiKeyPermission(str, Enum):
    """API密钥权限"""
    READ_ONLY = "read_only"
    READ_WRITE = "read_write"
    FULL = "full"


class ApiKeyStatus(str, Enum):
    """API密钥状态"""
    ACTIVE = "active"
    DISABLED = "disabled"
    EXPIRED = "expired"


class ApiKeyCreate(BaseModel):
    """创建API密钥"""
    key_name: str = Field(..., min_length=1, max_length=100, description="密钥名称")
    permission: ApiKeyPermission = Field(default=ApiKeyPermission.READ_WRITE, description="权限级别")
    ip_whitelist: Optional[List[str]] = Field(default=None, description="IP白名单")
    rate_limit: Optional[int] = Field(default=1000, ge=1, le=10000, description="每分钟请求限制")
    expires_days: Optional[int] = Field(default=None, ge=1, le=3650, description="有效期（天）")
    description: Optional[str] = Field(default=None, max_length=500, description="描述")


class ApiKeyUpdate(BaseModel):
    """更新API密钥"""
    key_name: Optional[str] = Field(None, min_length=1, max_length=100)
    permission: Optional[ApiKeyPermission] = None
    ip_whitelist: Optional[List[str]] = None
    rate_limit: Optional[int] = Field(None, ge=1, le=10000)
    status: Optional[ApiKeyStatus] = None
    description: Optional[str] = Field(None, max_length=500)


class ApiKeyResponse(BaseModel):
    """API密钥响应"""
    id: int
    account_id: int
    key_name: str
    api_key: str
    permission: ApiKeyPermission
    ip_whitelist: Optional[List[str]]
    rate_limit: int
    status: ApiKeyStatus
    usage_count: int
    last_used_at: Optional[datetime]
    last_used_ip: Optional[str]
    expires_at: Optional[datetime]
    description: Optional[str]
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class ApiKeyCreateResponse(ApiKeyResponse):
    """创建密钥响应（包含secret）"""
    api_secret: str  # 仅在创建时返回一次


class ApiKeyListResponse(BaseModel):
    """密钥列表响应"""
    total: int
    items: List[ApiKeyResponse]
    page: int
    page_size: int


class ApiKeyStats(BaseModel):
    """密钥统计"""
    total_keys: int
    active_keys: int
    disabled_keys: int
    expired_keys: int
    total_usage: int

File: frontend/src/api/apiKey.ts
import { request } from './index'

export interface ApiKey {
  id: number
  account_id: number
  key_name: string
  api_key: string
  permission: 'read_only' | 'read_write' | 'full'
  ip_whitelist: string[] | null
  rate_limit: number
  status: 'active' | 'disabled' | 'expired'
  usage_count: number
  last_used_at: string | null
  last_used_ip: string | null
  expires_at: string | null
  description: string | null
  created_at: string
  updated_at: string
}

export interface ApiKeyCreate {
  key_name: string
  permission?: 'read_only' | 'read_write' | 'full'
  ip_whitelist?: string[]
  rate_limit?: number
  expires_days?: number
  description?: string
}

export interface ApiKeyCreateResponse extends ApiKey {
  api_secret: string  // 仅创建时返回一次
}

// 获取密钥列表
export const getApiKeys = (params?: { page?: number; page_size?: number; status?: string }) => {
  return request.get('/api-keys', { params })
}

// 获取统计
export const getApiKeyStats = () => {
  return request.get('/api-keys/stats')
}

// 创建密钥
export const createApiKey = (data: ApiKeyCreate) => {
  return request.post<ApiKeyCreateResponse>('/api-keys', data)
}

// 更新密钥
export const updateApiKey = (id: number, data: Partial<ApiKeyCreate>) => {
  return request.put(`/api-keys/${id}`, data)
}

// 重新生成密钥
export const regenerateApiKey = (id: number) => {
  return request.post<ApiKeyCreateResponse>(`/api-keys/${id}/regenerate`)
}

// 删除密钥
export const deleteApiKey = (id: number) => {
  return request.delete(`/api-keys/${id}`)
}
